File: 0146-lru-cache/0146-lru-cache.cpp
class LRUCache {
    struct DBL{
        int key;
        int val; 
        DBL* next;
        DBL* prev;

        DBL(int key, int val){
            this->key = key;
            this->val = val;
            this->next = nullptr;
            this->prev = nullptr;
        }
    };
    unordered_map<int, DBL*> memo;
    DBL* lru;
    DBL* mru;
    int capacity;
public:
    LRUCache(int capacity) {
        this->capacity = capacity;
        lru = nullptr;
        mru = nullptr;
    }
    
    int get(int key) {
        if(memo.find(key) != memo.end()){
            if(memo[key]->next != nullptr){
                update(key);
            }
            return memo[key]->val;
        }
        return -1;
    }
    
    void put(int key, int value) {
        if(memo.find(key) != memo.end()){
            memo[key]->val = value;
            if(memo[key]->next != nullptr){
                update(key);
            }
            return;
        } 
        

        if(memo.size() == capacity){
            DBL* del = lru;
            lru = lru->next;
            del->next = nullptr;
            if(lru)
                lru->prev = nullptr;
            memo.erase(del->key);
            delete del;
        }

        if(memo.size() == 0){
            memo[key] = new DBL(key, value);
            lru = memo[key];
            mru = memo[key];
            return;
        }
        memo[key] = new DBL(key, value);
        memo[key]->next = nullptr;
        memo[key]->prev = mru;
        mru->next = memo[key];
        mru = memo[key];
    }

    void update(int key){
        if(!memo[key]->prev){
            mru->next = memo[key];
            mru = memo[key];
            lru = memo[key]->next;
            lru->prev = nullptr;
            return;
        }
        DBL* next = memo[key]->next;
        DBL* prev = memo[key]->prev;
        if(!prev){
            lru = next;
            mru = memo[key];
            return;
        }   
        if(prev){
            prev->next = next;
        }
        if(next){
            next->prev = prev;
        }
        mru = memo[key];
    }
};

/**
 * Your LRUCache object will be instantiated and called as such:
 * LRUCache* obj = new LRUCache(capacity);
 * int param_1 = obj->get(key);
 * obj->put(key,value);
 */

File: solutions/0146-lru-cache/0146-lru-cache.py
class DLL:

    def __init__(self, key, value, next = None, prev = None):
        self.key = key
        self.value = value 
        self.next = next
        self.prev = prev

class LRUCache:

    def __init__(self, capacity: int):
        self.capacity = capacity
        self.cache: dict[int, DLL] = {}
        self.head = DLL(-1, -1)
        self.tail = DLL(-1, -1)
        self.head.next = self.tail
        self.tail.prev = self.head

    def moveToHead(self, node: DLL, justInsert = False) -> None:
        if justInsert is False:
            node.prev.next = node.next
            node.next.prev = node.prev

        node.next = self.head.next
        self.head.next.prev = node
        node.prev = self.head
        self.head.next = node
    
    def remove(self, node: DLL) -> None:
        node.prev.next = node.next
        node.next.prev = node.prev
    
    def add_first(self, node: DLL) -> None:
        node.next = self.head.next
        self.head.next.prev = node
        node.prev = self.head
        self.head.next = node

    def get(self, key: int) -> int:
        if key not in self.cache:
            return -1
        
        self.remove(self.cache[key])
        self.add_first(self.cache[key])
        
        return self.cache[key].value

    def put(self, key: int, value: int) -> None:
        if key in self.cache:
            self.cache[key].value = value # update the value

            self.remove(self.cache[key])
            self.add_first(self.cache[key])
        else:
            node = DLL(key, value)
            self.cache[key] = node

            self.add_first(node)

        if len(self.cache) > self.capacity and self.tail.prev is not self.head:
            toDel = self.tail.prev
            self.remove(toDel)
            del self.cache[toDel.key]

            

# Your LRUCache object will be instantiated and called as such:
# obj = LRUCache(capacity)
# param_1 = obj.get(key)
# obj.put(key,value)    